### PR TITLE
[gcs] Fix GCS related issues: ByteSizeLong and redis connection

### PR DESCRIPTION
**Pull request by fishbone**
## Why are these changes needed?
- `ByteSizeLong` is expensive for our case, change it to other method (high CPU overhead)
- Re-use Redis connection in python layer. This reduces the number of redis connections (file descriptor overhead)

These changes together enable scaling GCS to 1000 nodes.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

Tested with 500 nodes x 20 actors

```
Ensuring actors have started: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 10000/10000 [01:57<00:00, 85.35it/s]
Actor time: 127.24993158900008
```
Redis connection is > 50k.
```
ubuntu@ip-172-31-22-58:~$ lsof -p 6470 -a | grep ESTABLISHED | wc -l
52263
```

### CODE REVIEW

**Comment by iycheng** [comment]
I also tested with 1k nodes x 20 actors:
```
Ensuring actors have started: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 20000/20000 [05:12<00:00, 64.08it/s]
Actor time: 331.57637213399994
```
Here is the script I run: https://gist.github.com/iycheng/d4fac42b339b135735e72cd985d50226
Here is the config I use: https://gist.github.com/iycheng/65293b834777cba072cd76a7e1e55c0b

@wuisawesome @ericl  could you verify that my setup is good?

File: src/ray/common/ray_config_def.h [hunk 1]
```diff
@@ -454,8 +454,8 @@ RAY_CONFIG(std::string, predefined_unit_instance_resources, "GPU")
 /// When set it to "FPGA", we will treat FPGA as unit_instance.
 RAY_CONFIG(std::string, custom_unit_instance_resources, "")
 
-// Maximum size of the batch size when broadcasting resources to raylet.
-RAY_CONFIG(uint64_t, resource_broadcast_batch_size_bytes, 1024 * 1024 * 5);
+// Maximum num of the batches when broadcasting resources to raylet.
+RAY_CONFIG(uint64_t, resource_broadcast_batch_num, 512);
```
Comment (ericl) [inline src/ray/common/ray_config_def.h:458]:
```suggestion
RAY_CONFIG(uint64_t, max_items_per_resource_broadcast, 512);
```

**Review by ericl — COMMENTED** [review]
LGTM but please rename to _size

File: src/ray/gcs/gcs_server/gcs_resource_manager.cc [hunk 2]
```diff
@@ -386,24 +387,26 @@ void GcsResourceManager::GetResourceUsageBatchForBroadcast(
     rpc::ResourceUsageBroadcastData &buffer) {
   absl::MutexLock guard(&resource_buffer_mutex_);
   resources_buffer_proto_.Swap(&buffer);
-  while (!resources_buffer_.empty() &&
-         buffer.ByteSizeLong() <
-             RayConfig::instance().resource_broadcast_batch_size_bytes()) {
-    auto element = std::begin(resources_buffer_);
-    buffer.add_batch()->mutable_data()->Swap(&element->second);
-    resources_buffer_.erase(element);
+  auto beg = resources_buffer_.begin();
```
Comment (rkooo567) [inline src/ray/gcs/gcs_server/gcs_resource_manager.cc:391]: This is a little unfortunate that we rely on cnt here... @wuisawesome is there any good way to get some heuristic size of each batch without calling that method?

**Review by ericl — APPROVED** [review]

File: src/ray/gcs/gcs_server/gcs_resource_manager.cc (hunk 2 above)
Comment (ericl) [reply src/ray/gcs/gcs_server/gcs_resource_manager.cc:391]: I prefer the size much more than getting the actual byte-size. It's simpler, and solves the same problem. The threshold can always be lowered if there is any remaining issues.

Comment (iycheng) [reply src/ray/gcs/gcs_server/gcs_resource_manager.cc:391]: It actually can be done in an estimated way, for example, get the first several ones and then do extrapolation.
I thought about that, but later I think maybe it's not a good idea. 1) It's a little bit hard to understand how the num of RPC requests are sent. For example, with X nodes, we don't have a good estimation of how many RPCs we need for pull-push. Another one is the SerDe is too expensive for these messages, and we need to just bypass them in some way. For example, right now, the SerDe is O(N^2) [N=# of nodes], but actually, we can make it O(N), and we can cap later after optimization.

**Pull request merged**